src/warc2zim/url_rewriting.py

```diff
@@ -99,49 +99,186 @@
     },
 ]
 
+
 COMPILED_FUZZY_RULES = [
     {"match": re.compile(rule["pattern"]), "replace": rule["replace"]}
     for rule in FUZZY_RULES
 ]
 
 
-def reduce(path: str) -> str:
-    """Reduce a path"""
+class HttpUrl:
```

Comment:
I still would like that `str(HttpUrl("http://whatever"))` returns `"HttpUrl(http://whatever)"` for easier debugging, so I still need the custom `value` property and use this property wherever needed. Does it make sense or I'm just abusing the `__str__` meaning? I could have used `__repr__` instead of `value` but I find this even more confusing.

Reply:
Plus one drawback of subclassing is that I do not get a pyright error when I forget to use `.value` property ; I don't think that subclassing is a good idea in fact.

Reply:
>  still would like that str(HttpUrl("http://whatever")) returns "HttpUrl(http://whatever)"

this is exactly what `repr` is for! (hence the name)

```py
class HttpUrl(str):
    def __new__(cls, content):
        HttpUrl.check_validity(content)
        return str.__new__(cls, content)

    def __repr__(self) -> str:
        return f"{type(self).__name__}({super().__repr__()})"

    @classmethod
    def check_validity(cls, value: str) -> None:
        ...

repr(HttpUrl("AAA"))
> "HttpUrl('AAA')"
```

> I do not get a pyright error when I forget to use .value property

The whole point is to not use `.value`. If you expect an `HttpUrl` and you pass it an `str`, pyright will complain